### ISSUE TITLE
Math::BigInt::LTM breaks on more complex example

### ISSUE THREAD

**Issue by brtastic** [DCIT/perl-CryptX#74]
Hello,
I'm sorry if this report is not very clear. I'm pretty sure I found a bug in Math::BigInt::LTM, but I'm having problems reproducing it on a simple example. I seem to have the newest version of the package (0.073) as well as the newest Math::BigInt.
The problem is I get one of those errors:
```
Math::BigInt::LTM::_mul: y is not of type Math::BigInt::LTM at .../perl5/Math/BigInt.pm line 1760.

Math::BigInt::LTM::_acmp: n is not of type Math::BigInt::LTM at .../perl5/Math/BigInt.pm line 1365
```

Basically what I got is a Moose class here: https://github.com/brtastic/btcpaywall/blob/master/lib/Model/Request.pm
Which is constructed right here: https://github.com/brtastic/btcpaywall/blob/master/lib/BtcPaywall/Form/Request.pm#L82

The error occurs in the constructor call, so my code is not touched at all, its just Moose using my class configuration. The first time the interpreter encounters this code, it works as it should. Then it proceeds to do more advanced stuff, and it loads Math::BigInt with LTM in the process.
The second time the interpreter encounters this code, I get the first error shown above (with _mul). I have no idea why - the class hasn't really got heavy arythmetic in its properties. I tried to stacktrace the error with Carp::Always, but it doesn't seem to work.
If I load Math::BigInt with LTM at the top of this file I get that error in the first run as well. That got me to believe it was Math::BigInt::LTM problem. I had two places that used LTM, and changing only one of those produced the second error listed above (with _acmp). After replacing LTM with GMP in both places the problem has gone away.

I know this is vague, but it's all I got for now, as my attempts to reproduce with simple examples failed. Hopefully it is enough information for you to fix this, but if you want me to check something on the environment I found this on, I'll be happy to.

**Comment by brtastic** [comment]
It might be connected to some recent changes in Math::BigInt, because version 1.999822 was breaking even GMP, as I reported here: https://rt.cpan.org/Public/Bug/Display.html?id=137491. However, I am now using the patched 1.999823, and the bug still occurs.

**Comment by karel-m** [comment]
@pjacklam do you have any hints where to look?

**Comment by pjacklam** [comment]
> It might be connected to some recent changes in Math::BigInt, because version 1.999822 was breaking even GMP, as I reported here: https://rt.cpan.org/Public/Bug/Display.html?id=137491. However, I am now using the patched 1.999823, and the bug still occurs.

Math-BigInt 1.999822 didn't break Math::BigInt::GMP or any other of the backend libraries. The backend libraries are completely independent of Math::BigInt, Math::BigFloat, Math::BigRat and other modules using these backend libraries, so Math::BigInt etc. couldn't break these backend modules even if they tried. The dependency goes the other way: Math::BigInt etc. are dependent on the backend libraries.

The "breakage" you are referring to are some failed tests in the test suit for the Math::BigInt::GMP backend library. Almost all of the tests in the test suits for the backend libraries don't actually test the backend library directly, but rather test how Math::BigInt etc. work with the given backend library. So any change in the behaviour Math::BigInt is likely to break some tests in the test suits for the backend libraries. For example, if I change Math::BigInt so that Math::BigInt->new("") returns 0 rather than NaN, the test suit for every backend library will fail.

**Comment by pjacklam** [comment]
> @pjacklam do you have any hints where to look?

I have looked into this, but it is not clear to me what goes wrong. It would be very helpful if I could reproduce the problem.

It seems to be important that Math::BigInt et al. catch the case when more than one backend library is loaded.

**Comment by brtastic** [comment]
I'll try once again to reproduce the error with a simpler case, but it has to wait until the weekend

**Comment by brtastic** [comment]
I found some time before the weekend, and I now understand the problem. It is indeed caused by mixed backends. Can be reproduced with:
```
use Data::ULID;
use Math::BigInt try => 'LTM,GMP';

Data::ULID::ulid();
```
Data::ULID has this in its code:
```
use Math::BigInt lib => 'GMP', upgrade => 'Math::BigFloat';
```
and Bitcoin::Crypto has this:
```
use Math::BigInt 1.999808 try => 'LTM,GMP';
```
I use both in this project, which is why I get this error. However I still don't understand if this is an error with one of the libraries, or just the way Math::BigInt works in general.

**Comment by pjacklam** [comment]
Thank you for the example. What you see is not a problem with the libraries, but rather a limitation of Math::BigInt. Math::BigInt has never supported mixing libraries.

It is certainly possible to add support for mixing libraries. Every instance should have a field telling which library is used by that instance. Math::BigInt will then have to check which libraries are used by each operand, and if the libraries differ, then Math::BigInt will have to do something like

```
$y = $LIB2 -> _new($LIB1 -> _str($x));
```
So it is certainly possible. If the libraries are identical, the overhead should be minimal, but if they differ, the conversion will certainly introduce some overhead. On the other hand, it will make the code work, which might make it worth it.

**Comment by brtastic** [comment]
Thanks for the insights. So right now, without the fix you're describing, what action can I take to fight this bug? What I've done was replace the order in which Bitcoin::Crypto tries to load libraries directly in its code, so it loads GMP as its first choice. However it seems like a programmer currently has very few options if two libraries are incompatible when it comes to Math::BigInt backends, as simply loading a different backend in one package will cause it to change how it works in another package. Currently the problem seems to be occuring due to constants, but it might also be due to object data or class lexical variables (any Math::BigInt instance saved in a persistent variable). It would be nice if I could load Math::BigInt with lets say GMP at the start of my script and tell it to stick to that library. I guess that's currently impossible?

Also, sorry for opening the ticket in CryptX, it really seemed like LTM issue to me at first.

**Comment by pjacklam** [comment]
> Thanks for the insights.

Thank you! I didn't realize until now how real the problem is with multiple module authors explicitly requesting different backend libraries for Math::BigInt.

> So right now, without the fix you're describing, what action can I take to address this behavior?

At the moment, I'm not sure there is that much you can do other than hacking the modules you need.

> It would be nice if I could load Math::BigInt with lets say GMP at the start of my script and tell it to stick to that library. I guess that's currently impossible?

I can make Math::BigInt die with a meaningful error message, if more than one library is used. While that will be more informative than the current behaviour, it won't really solve your problem.

I can make Math::BigInt use the first library loaded, and ignore any subsequent attempts at loading a different library. That should work, unless there is something I haven't thought of.

> Also, sorry for opening the ticket in CryptX, it really seemed like LTM issue to me at first.

I guess this should be a Math::BigInt ticket, but I fully understand that it was not obvious what was the source of the issue you were experiencing.

**Comment by brtastic** [comment]
Luckily I am the author of the Bitcoin library that loads LTM, so I'll switch the order in which it tries to load backends, which will make more sense anyway (since it always gets LTM due to dependencies). This should be addressed somehow anyway, but I obvoiusly don't know ins and outs of the solutions you mentioned. To me, respecting the previous choice of backend plus a more informative error message would do, as it would take the power to decide back to the module user.

Let me close this ticket, and I'll open a new one on Math::BigInt bugtracker later today. Thank you very much for helping me investigate this.

**Comment by pjacklam** [comment]
This issue should be fixed in the newly released Math-BigInt-1.999824. Now it works like this: The first backend library that is successfully loaded is used. Any further attempts at loading a different library are ignored.
